portability: Towards C99/C11 compliance

mped-oticon opened the pull request:
This lessens the use of GNU extensions and permits code to compile with our Clang-based compiler.
Issues still remain, but these may require discussion; hopefully, the changes herein are acceptable. Code readability remains unharmed :)

Issue #9879.

mped-oticon committed:
portability: Remove C99-illegal semicolons from the global scope

This commit touches the C codebase and the python syscall generator.

The Z_GENLIST-macros expand to whole functions. Once expanded by the
preprocessor we notice a semicolon is put after the function body. But
ISO C99 does not allow extra ‘;’ outside of a function. Though this is
accepted by GCC with GNU extensions, it is not by Clang.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
portability: Ensure no C99-illegal semicolons exists in structs

Macro _OBJECT_TRACING_NEXT_PTR expands to a member or to nothing.
Macro _OBJECT_TRACING_NEXT_PTR is used in a number of places, like:

        struct k_stack {
                .. omitted ..
                _OBJECT_TRACING_NEXT_PTR(k_stack);
                u8_t flags;
        };

When the macro expands to nothing, a lonesome semi would remain. This is
illegal in C99, but permitted in GCC with GNU extensions.

Rather than expand to empty, we now expand to a zero-length array.
This means we can retain the trailing semis across structs wherein the
macro is used.

Note that zero-length array (foo[0]) != flexible array member (foo[]):
 * zero-length array: Is GNU+Clang extension. Anywhere in struct.
 * flexible array member: Is C99. Only in end of struct.

Thus we have really only traded-off one portability issue for
another, more acceptable, one at least.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -264,6 +264,9 @@ zephyr_system_include_directories(${NOSTDINC})
 zephyr_cc_option(-Werror=implicit-int)
 
+# Prohibit void pointer arithmetic. Illegal in C99
+zephyr_cc_option(-Wpointer-arith)
```
-Werror=pointer-arith ?

Unless there is a good reason to allow pointer arithmetic during development ...

mped-oticon replied:
Too aggressive? I did consider it but may lead to widespread breakage for some boards accustomed to GCC.

SebastianBoe replied:
Warnings turn into errors in CI so the breakage should be the same.

pfalcon replied:
Ideally, this should have been a separate commit. So, fixing true source-code errors was separated from debatable build system config change.

mped-oticon replied:
@pfalcon Agree. Maybe even another, later, PR?

pfalcon replied:
> Agree.

Thanks.

> Maybe even another, later, PR?

Whatever is easier for you. Just my 2 cents, I'd prefer it to be warning indeed, not error right away. As @SebastianBoe mentioned, warning will turn to errors during CI, so bad code won't leak into mainline by overlook. But during development/testing/debugging, sometimes it's helpful to take hacky shortcuts, so having it just as a warning is helpful.

jhedberg commented on subsys/bluetooth/host/hci_ecc.c:
```diff
@@ -223,7 +223,7 @@ static void clear_ecc_events(struct net_buf *buf)
 	struct bt_hci_cp_le_set_event_mask *cmd;
 
-	cmd = (void *)buf->data  + sizeof(struct bt_hci_cmd_hdr);
+	cmd = (void *)((u8_t *)buf->data  + sizeof(struct bt_hci_cmd_hdr));
```
The `buf->data` is already `u8_t *` so this particular cast looks redundant. The added parenthesis behind the pointer increment is still needed as a fix though (I think).

pfalcon replied:
+1

jhedberg reviewed (changes requested):
Looks generally good. One small comment regarding a redundant typecast.

andyross commented on include/logging/log_backend.h:
```diff
@@ -147,7 +147,7 @@ static inline const struct log_backend *log_backend_get(u32_t idx)
 static inline int log_backend_count_get(void)
 {
-	return ((void *)__log_backends_end - (void *)__log_backends_start) /
+	return ((u8_t *)__log_backends_end - (u8_t *)__log_backends_start) /
```
Likely to be a huge bike shed, but personally I'd prefer "char" for the type to use here, which is guaranteed by the standard to be the minimal memory unit size and readable to all.  Our oddball unit types are for unifying size (e.g. long vs. int) and signedness (e.g. char-is-unsigned-on-arm-wtf?).

But others feel probably even more strongly in the other direction, so just a note.

ulfalizer replied:
*Backseat driving*
What's with the `u8_t`, etc., types in Zephyr btw? Does Zephyr need to support pre-C99 compilers without `<stdint.h>`?

ulfalizer replied:
It's a mish-mash in the Linux kernel too, but for something like Zephyr it might be reasonable to do a global search-replace I guess, unless there's reasons to avoid the `<stdint.h>` types.

andyross replied:
We actually did use the standard types pervasively, and the existing types are the **result** of a global search and replace.  :)

I mostly stayed out of the fight, but yeah: inconsistent toolchain support for stdint.h was part of the reasoning.

mped-oticon replied:
If char-is-unsigned-on-arm-wtf, then I would prefer u8_t cause I know it won't change signedness under my feet -- and it should, knock-on-wood, always be unsigned 8 bit for all supported platforms. Also for consistency.

ceolin replied:
I rather prefer using stdint, but if we use our types, u8_t is better than char, because it explicitly says that is is unsigned. Moreover, this is also a MISRA-C recommendation, if we want be compliant, we need to move in this direction.

mped-oticon replied:
Leaving as-is.

andyross commented on include/misc/sflist.h:
```diff
@@ -314,7 +314,7 @@ static inline void sys_sfnode_flags_set(sys_sfnode_t *node, u8_t flags)
 static inline bool sys_sflist_is_empty(sys_sflist_t *list);
 
-Z_GENLIST_IS_EMPTY(sflist);
+Z_GENLIST_IS_EMPTY(sflist)
```
Removing all those semis from the top level is just asking to blow up our poor editors' indentation logic.  But I guess these are code-generation macros, so...

mped-oticon replied:
Macro hackery always deserves some punishment ;-)

andyross reviewed (approved):
Not loving some of the aesthetics, but the issues here are unambiguous.  I still don't understand why a "clang-based" compiler is going to have trouble with this, unless it's a really really old version of clang.

mped-oticon commented:
BTW, is there a way to invoke CI/shippable before spamming y'all?

pfalcon commented:
@mped-oticon :

> BTW, is there a way to invoke CI/shippable before spamming y'all?

See previous comment.

pfalcon commented:
(Maybe you meant something else. In this case, see scripts/{checkpatch.pl, sanitycheck}, etc.)

pfalcon commented on include/kernel.h:
```diff
@@ -95,7 +95,7 @@ typedef struct {
 
 #ifdef CONFIG_OBJECT_TRACING
-#define _OBJECT_TRACING_NEXT_PTR(type) struct type *__next
+#define _OBJECT_TRACING_NEXT_PTR(type) struct type *__next;
```
I'm not sure what happens here. You add a semicolon here to remove it in bunch of places and get the same result?

ceolin replied:
yeah, this commit makes no sense at all for me

pfalcon replied:
> It seems it's to avoid some warnings of dangling semicolons for the condition when the macro expands to nothing. 

Ah, d'oh. Too bad that we have 2 threads discussing the same line of code...

pfalcon reviewed (changes requested):
Please confirm what happens here.

pfalcon commented on scripts/gen_syscalls.py:
```diff
@@ -147,7 +147,7 @@ def analyze_fn(match_group):
     argslist = ", ".join(flat_args)
 
-    invocation = "%s(%s);" % (macro, argslist)
+    invocation = "%s(%s)" % (macro, argslist)
```
Please add to commit message explicit mention of what is done here. E.g.

> Also update scripts/gen_syscalls.py to not generate extra semicolons after K_SYSCALL_DECLARE* macros.


Again, if this was my project, I would request to split this commit in 2, both with focused commit messages, as it updates 2 widely different parts of the codebase.

pfalcon reviewed (changes requested):
More detailed commit messages needed.

jukkar commented on include/kernel.h:
```diff
@@ -1431,7 +1431,7 @@ struct k_timer {
 	void *user_data;
 
-	_OBJECT_TRACING_NEXT_PTR(k_timer);
+	_OBJECT_TRACING_NEXT_PTR(k_timer)
```
It would probably be more clearer, if the commit message would explain why this change is needed. Now it just states what is being done but that is quite obvious by looking the code. The real issue is that if the macro is not defined, then there is extra semicolon inside the struct.

nashif replied:
isnt that in the commit message already?

`ISO C99 does not allow extra ‘;’ outside of a function`

jhedberg replied:
Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition be avoided by having it expand to a dummy zero-length array element instead of nothing?

pfalcon replied:
> isnt that in the commit message already?

> ISO C99 does not allow extra ‘;’ outside of a function

That commit message is possibly wrong (doesn't tell the whole truth).

pfalcon replied:
> Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition

What's so confusing and unusual? In normal languages, macros are Turing-complete, imagine how that is "confusing". And in C, I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

pfalcon replied:
Ok, @jhedberg explains the issue here: https://github.com/zephyrproject-rtos/zephyr/pull/9917#discussion_r217603899 (too bad we have 2 branches of discussion of the same issue).

Which only shows that there can't be too detailed commit message, and majority of commit messages are terse up to confusion and doesn't described the changes they do well.

jhedberg replied:
> I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

@pfalcon I didn't say that it's disallowed. It's just that most macros in the code base don't do this, which to the casual reader makes it look like there's a bug in the code that's using the macro (since one expects to see a semicolon at the end of each struct member declaration). Anyway, I'm curious to hear thoughts on my zero-length array proposal that should help avoid all this.

pfalcon replied:
> with which compiler issues fixed in this PR were discovered?

Ok, the description says "permits code to compile with our Clang-based compiler.". Is that public compiler? Based on recent Clang?

Anyway, I guess it's up to @mped-oticon to look at zero-sized array solution and test if it works/see if he'd be willing to use it.

(My concern with this PR overall remains not detailed enough commit messages.)

mped-oticon replied:
@jhedberg zero-length array (aka flexible array member) is a fine idea/hack but has caveats: Look at `kernel.h`'s `struct k_stack` (or k_msgq or k_pipe), is the last member is some `u8_t flags`. That means our zero-length array is not the last member. This violates C99, see [1].

I would also be very skeptical of nesting structs that contain flexible array members - as the address for subsequent members would not be known at compile-time (which is required). I don't know if these struct nest though, but even if not, it could blow up later.

[1]
C99 spec: "As a special case, the **_last_** element of a structure with more than one named member may have an incomplete array type; this is called a flexible array member."

mped-oticon replied:
@pfalcon 
> [...] looks confusing IMHO. So, doing reasonable steps to preserve them makes sense IMHO.
Hopefully the ALL_CAPS and parenthesis should inform you that this line is a macro, and therefore caution should be exercised. I agree it would be nice to keep up appearence, but IMHO the benefit (portability) outweighs the cons (readabillity). A popular RTOS targetting embedded systems will have ugly corners. Good thing we're not in a beauty contest :)

mped-oticon replied:
@pfalcon I will add make the commit messages more detailed.

jhedberg replied:
@mped-oticon I wasn't aware that this wasn't formally allowed elsewhere than at the end of structs. The main reason I thought it was is that there's a major project that uses it, namely the Linux kernel:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n737
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n750
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n761

There are probably many more occurrences, but I pointed out these from the subsystem that I'm most familiar with. Anyway, because Linux uses zero-length arrays like this, it would seem bad business for a compiler not to support it, so I wonder how likely it is that we'd run into issues.

pfalcon replied:
>  I will make the commit messages more detailed

Thanks, appreciated!

mped-oticon replied:
@pfalcon, @jhedberg It may work out with [0], but [] definitely does not. I thought zero-sized array and flexible arrays were treated the same, but I guess not :)

https://gcc.godbolt.org/z/1WEeEq 
removing `-pedantic` quenches the warning. So I guess [0] is fine if we never want `-pedantic` with `-Wall`, which would be a PITA.

That leaves me satisfied with the [0] proposal.

nashif commented:
regarding the ‘;’, I guess this is what you get when you build with -pedantic, I have seen this also with clang..

```
                                                                      ^
/home/nashif/github/zephyr/build/zephyr/include/generated/syscalls/kernel.h:69:87: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
 K_SYSCALL_DECLARE1_VOID(K_SYSCALL_K_QUEUE_INIT, k_queue_init, struct k_queue *, queue);
                                                                                       ^
/home/nashif/github/zephyr/build/zephyr/include/generated/syscalls/kernel.h:71:101: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
 K_SYSCALL_DECLARE1_VOID(K_SYSCALL_K_QUEUE_CANCEL_WAIT, k_queue_cancel_wait, struct k_queue *, queue);
                                                                                                     ^

```

mped-oticon committed:
portability: Avoid void* arithmetics which is a GNU extension

Under GNU C, sizeof(void) = 1. This commit merely makes it explicit u8.

Pointer arithmetics over void types is:
 * A GNU C extension
 * Not supported by Clang
 * Illegal across all ISO C standards

See also: https://gcc.gnu.org/onlinedocs/gcc/Pointer-Arith.html

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
portability: Warn at void* arithmetics which is a GNU C extension

Ensure future commits do not re-introduce use of this GNU C extension.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

andyross commented:
Just to dab a tiny bit more paint on the shed:  is this really just a pedantic-level warning from clang?  @mped-oticon, can you detail exactly what your compiler's misfeatures are?  If it's just an issue with warning flags vs. very-long-established-industry-convention-in-all-other-projects, maybe we don't need to care?

nashif commented:
> Just to dab a tiny bit more paint on the shed: is this really just a pedantic-level warning from clang?

the above is from gcc.

The reason why we should care about standards, compilers behave differently, big difference between gcc and clang and the default behaviour, for one it is pedantic, for the other it is a warning that we would need to suppress, some compilers might just reduce to build the code, maybe not in this case, but I would try as much as possible be consistent and do not spend too much time thinking of how to solve it in zephyr build system working around the compilers.

The argument now is that it works as is so why bother, the counter argument is that someone will try some commercial compiler or some new version and things will break, we noticed that when we started build with llvm few years back.

mped-oticon commented:
> @mped-oticon, can you detail exactly what your compiler's misfeatures are? 
Yes: We don't have support for GNU extensions in our Clang.

That means you will get this list incrementally, as each PR will contain what we need and not more.
We have already been exploring compiler options, permissive freestanding, etc, but without this PR, we get errors, not just warnings.

I'd be happy to elaborate more on our challenges with our current toolchain, but that would quickly become a long text :)

ceolin commented on lib/libc/minimal/source/string/string.c:
```diff
@@ -203,7 +203,7 @@ void *memmove(void *d, const void *s, size_t n)
 	const char *src  = s;
 
-	if ((size_t) (d - s) < n) {
+	if ((size_t) (dest - src) < n) {
```
ok, we may have a problem here, shouldn't you cast to `ssize_t` to avoid an underflow ? Instead of using `char *` it may be better using something `uint8_t` from `stdint.h` that is part of C standard and is recommended by other code guidelines.

mped-oticon replied:
> ok, we may have a problem here, shouldn't you cast to ssize_t to avoid an underflow ? 

Nope. I tried to make an ASCII drawing of why size_t is correct:
```
A):
    |            ssssssssssss  
    |      dddddddddddd         must be written forwards

                if ((size_t) (0 - 10) < 12) {
                if ((size_t) (-10) < 12) {
                if ((size_t) (-10) < 12) {
                if (4294967286 < 12) {      or some other big number
                if (false) {    False, so we do forward writes

B):
    |ssssssssssss              
    |      dddddddddddd         must be written backwards
 
                if ((size_t) (dest - src) < n) {
                if ((size_t) (10 - 0) < 12) {
                if ((size_t) 10 < 12) {
                if (10 < 12) {
                if (true) {     True, so we do backwards writes

C):
    |      dddddddddddd        dont care, could be no-op
    |      ssssssssssss
```

mped-oticon replied:
> Instead of using char * it may be better using something uint8_t from stdint.h that is part of C standard and is recommended by other code guidelines.

I'm confortable with char*, but agree uint8_t* would be better. However I also know of  platforms having problems with stdint.h.

ceolin commented on lib/mempool/mempool.c:
```diff
@@ -73,14 +73,14 @@ static size_t buf_size(struct sys_mem_pool_base *p)
 	int i;
 	size_t buflen = p->n_max * p->max_sz, sz = p->max_sz;
-	u32_t *bits = p->buf + buflen;
+	u32_t *bits = (u32_t *)((u8_t *)p->buf + buflen);
```
can't you use `uintptr_t` here ? It's more portable that `u32_t`

mped-oticon replied:
I don't see why. It has to become a pointer-to-u32_t regardless.
In this translation unit, u32_t is fine from a portability perspective as Zephyr promises to make that type available.

ceolin commented on subsys/fs/nvs/nvs.c:
```diff
@@ -149,7 +150,7 @@ static int _nvs_flash_block_cmp(struct nvs_fs *fs, u32_t addr, const void *data,
 		len -= bytes_to_cmp;
 		addr += bytes_to_cmp;
-		data += bytes_to_cmp;
+		data = (const u8_t *)data + bytes_to_cmp;
```
shouldn't use `data8` here too ?

mped-oticon replied:
Good catch. I'll fix to a consistent style.

ceolin reviewed (commented):
I'm 100% in favor of being compliant with C standard. It will make the code much more portable.

superna9999 commented:
ok for `drivers/flash/flash_stm32.c`

pfalcon commented on ext/hal/libmetal/libmetal/lib/io.c:
```diff
@@ -33,7 +33,7 @@ void metal_io_init(struct metal_io_region *io, void *virt,
 	       void *restrict dst, int len)
 {
-	void *ptr = metal_io_virt(io, offset);
+	unsigned char *ptr = (unsigned char *)metal_io_virt(io, offset);
```
@mped-oticon, this makes changes to ext/hal/ , and that's usually a good reason to make that a separate commit, and review would usually include a question "did you submit these changes upstream"?

@galak: FYI

mped-oticon replied:
@pfalcon See mine and Carles' comment above.
Should I update our copy of libmetal (where this has been fixed -- in a slightly different way than this PR)?

pfalcon replied:
> Should I update our copy of libmetal (where this has been fixed -- in a slightly different way than this PR)?

Sure, that's the best solution, please coordinate with @galak who maintains Zephyr's copy. Also cc: @wjliang, upstream maintainer.

pfalcon reviewed (approved):
Thanks for all the refactors and detailed commit messages! +1 from me, but I found another concern: patching ext/hal/, which may need to be propagated upstream. Sorry if that causes additional delays, I'd personally like this merged asap!

pfalcon commented:
@jhedberg : The cause of your -1 was fixed, can you re-review?

mped-oticon commented on ext/hal/libmetal/libmetal/lib/io.c:
```diff
@@ -33,7 +33,7 @@ void metal_io_init(struct metal_io_region *io, void *virt,
 	       void *restrict dst, int len)
 {
-	void *ptr = metal_io_virt(io, offset);
+	unsigned char *ptr = (unsigned char *)metal_io_virt(io, offset);
```
@carlescufi It appears already fixed in libmetal: https://github.com/OpenAMP/libmetal/issues/65
Though, Zephyr's copy is not up to date.

carlescufi replied:
Right, that's why we need @galak to import a new version. I believe he's already working towards that.

mped-oticon commented on ext/fs/nffs/src/nffs_write.c:
```diff
@@ -353,7 +353,7 @@ nffs_write_chunk(struct nffs_inode_entry *inode_entry, uint32_t file_offset,
         data_offset = cache_block->ncb_file_offset + chunk_off - file_offset;
         rc = nffs_write_over_block(cache_block->ncb_block.nb_hash_entry,
-                                   chunk_off, data + data_offset, chunk_sz);
+                                   chunk_off, (const uint8_t *)data + data_offset, chunk_sz);
```
@carlescufi Thanks, I overlooked that one. Will do.

mped-oticon replied:
@carlescufi PR created: https://github.com/apache/mynewt-nffs/pull/11
So I guess the process will be: Get mynewt-nffs PR merged, import fresh mynewt-nffs in Zephyr, rebase this PR. Correct?

carlescufi replied:
Correct. See the commit log in `ext/` to look at the format of commit messages that import external code.

mped-oticon commented:
We need to reimport libmetal (should be OK) and mynewt-nffs (pending PR).
Should I do this in a separate Zephyr PR?

carlescufi commented:
@mped-oticon 

>Should I do this in a separate Zephyr PR?

It's fine if it's this PR, but separate commits.

mped-oticon committed:
ext: nffs: portability: Avoid void* arithmetics which is a GNU extension

Under GNU C, sizeof(void) = 1. This commit simply uses uint8_t.

Pointer arithmetics over void types is:
 * A GNU C extension
 * Not supported by Clang
 * Illegal across all ISO C standards

See also: https://gcc.gnu.org/onlinedocs/gcc/Pointer-Arith.html

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
ext: libmetal: Update import of libmetal from b4b5bea to a4f7630

Pull in a more recent version of libmetal, motivated by the void
pointer fixes in efc33fe.

libmetal's b4b5bea to a4f7630 spans:

a4f7630 linux: device: fix max chars to copy for the dev name
10a0d5b lib: system: freertos: mutex: change to use atomic_int ...
4e670c5 ci: Zephyr env var for toolchain
3bef6f0 cmake: Fix Zephyr library integration
13158c3 compiler: add IAR compiler file
0820a2e microblaze: suppress io 64-bit R/W access if 64 bit atomic ...
e2e5608 io: suppress io 64 bit R/W access if 64 bit atomic not ...
efc33fe io: fix compilation error for void pointers.
32ad1d2 cache: fix compilation issue
c1ade0d travis CI: update Zephyr SDK version to 0.9.3
c49d7cd travis CI: updated arm compiler ubuntu package
d279a6a travis ci: Update FreeRTOS download path
5c70105 irq: update sys_irq_restore_enable and sys_irq_save_disable
05f0cd4 cmake: Only look for doxygen package if docs enabled
b4b5bea zephyr: Introduce WITH_ZEPHYR_LIB option

At current time of writing, libmetal's HEAD is 5bc2106641.
However this failed to build in their Travis CI.
The latest passing commit is a4f7630.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>
